### PR TITLE
EQL: correct time accounting for an intermediary response

### LINKED ISSUE #75825: EQL: intermediary async responses reported as taking negative time
When an async search goes over the specified sync waiting time (`wait_for_completion_timeout`), the time reported to provide the answer (`took`) is negative:
```
    "id": "FjBycjdDTXRDVExPOHdjTUkzWUVCV2cbSlJUTExNUldRTHVKTlhieHFvVVlyUToyMTIx",
    "is_partial": true,
    "is_running": true,
    "took": -1627484284881,
    ...
```

Expected is for the response time to be non-negative.

### PR DESCRIPTION
**Pull request by bpintea**
This corrects the accounting of answer time reported in "took" attribute
of an intermediary EQL async response. This has been reported as a
negative value, since the delta mixes wall-clock time with VM time.

(The confusion is favoured by Task#getStartTime() returning a wall-clock
time, while Task#getStartTimeNanos() VM's time.)

Closes #75825.

### CODE REVIEW

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java [hunk 1]
```diff
@@ -95,7 +95,8 @@ public void execute(EqlSearchRequest request, EqlSearchTask task, ActionListener
     @Override
     public EqlSearchResponse initialResponse(EqlSearchTask task) {
         return new EqlSearchResponse(EqlSearchResponse.Hits.EMPTY,
-            threadPool.relativeTimeInMillis() - task.getStartTime(), false, task.getExecutionId().getEncoded(), true, true);
+            TimeValue.nsecToMSec(task.getStartTimeNanos()) - threadPool.relativeTimeInMillis(), false, task.getExecutionId().getEncoded(),
```
Comment (imotov) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java:98]: getStartTimeNanos is obtained from System.nanoTime(). So, would something like `TimeValue.nsecToMSec(System.nanoTime() - task.getStartTimeNanos())` make more sense in this case?

Comment (bpintea) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java]: Thanks, indeed, _that_ makes sense.

**Comment by bpintea** [comment]
@elasticmachine update branch

**Review by Luegg — APPROVED** [review]
lgtm 👍 

Just wondering whether we have a general preference for the JVM time? Otherwise, `System.currentTimeMillis() - task.getStartTime()` might also be an option?

**Comment by bpintea** [comment]
> Just wondering whether we have a general preference for the JVM time? Otherwise, System.currentTimeMillis() - task.getStartTime() might also be an option?

You're right, the original code was doing millis math in the first place. I guess nanos math makes sense to more accurately differentiate between 0 and 1ms, but don't think this is that relevant in this case. Applied your suggestion.

**Review by astefan — APPROVED** [review]
LGTM
This is fixing an annoying issue that I noticed myself as well. Thanks for addressing it.

**Comment by bpintea** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by imotov** [comment]
> Just wondering whether we have a general preference for the JVM time?

I don't think it is a general preference, but for calculating elapsed time we are trying to go with `System.nanoTime()` nowadays. There are a couple of reasons for that. `System.nanoTime()`  is monotonic (which exception of some [JVM bugs](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6458294)), while `System.currentTimeMillis()` is not by design. In other words, `System.currentTimeMillis() - task.getStartTime()` can be negative if time on the system clock is update by a user or an NTP server at just the right moment. On some system you can also expect `System.nanoTime()` to be more precise than `System.currentTimeMillis()`. I saw reports of the first having granularity of hundreds of nanoseconds, while the second only ticking every 10ms or even 50ms on some older Windows systems. One notable exception when we would want to use `System.currentTimeMillis()` for elapse time calculation is if we move the data between nodes. `System.nanoTime()` doesn't really mean anything outside of JVM where the start time was computed, so if you start an event on one node and finish on another, `System.currentTimeMillis()` has to be used (but then all bets regarding monotonicity are off anyway since we don't synchronize clocks).

**Comment by bpintea** [comment]
@elasticmachine update branch

**Comment by bpintea** [comment]
@elasticmachine run elasticsearch-ci/docs

**Comment by bpintea** [comment]
Thanks for the explanations, @imotov, reverted to your proposal.

**Pull request merged**